File: task_equivalents/llm_translations/gpt/optimized/python/machine_learning_workflows/02_regression_workflow_feature_importance.py
import numpy as np
import pandas as pd
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.linear_model import LassoCV
from sklearn.metrics import mean_squared_error

hitters = pd.read_csv("data/csv/Hitters.csv", index_col=0).dropna()

drop_cols = [col for col in ["rownames", "Player", "Name"] if col in hitters.columns]
hitters = hitters.drop(columns=drop_cols)

hitters["League"] = (hitters["League"] != "A").astype(int)
hitters["NewLeague"] = (hitters["NewLeague"] != "A").astype(int)
hitters["Division"] = (hitters["Division"] != "E").astype(int)
hitters["logSal"] = np.log(hitters["Salary"])

x = hitters.drop(columns=["Salary", "logSal"])
y = hitters["logSal"]

rng = np.random.RandomState(1)
train = rng.choice(len(x), size=200, replace=False)
test = np.setdiff1d(np.arange(len(x)), train)

xtrain = x.iloc[train]
xtest = x.iloc[test]
ytrain = y.iloc[train]
ytest = y.iloc[test]

train_mse = np.empty(20)
lambda_grid = np.linspace(0.001, 0.201, 20)

for i, lam in enumerate(lambda_grid):
    boost_hit = GradientBoostingRegressor(
        loss="squared_error",
        n_estimators=1000,
        learning_rate=lam,
        random_state=1,
    )
    boost_hit.fit(xtrain, ytrain)
    boost_pred = boost_hit.predict(xtrain)
    train_mse[i] = mean_squared_error(ytrain, boost_pred)

cv_out = LassoCV(cv=10, random_state=1).fit(xtrain.to_numpy(), ytrain.to_numpy())
bestlam = cv_out.alpha_
lasso_pred = cv_out.predict(xtest.to_numpy())
lasso_mse = mean_squared_error(ytest, lasso_pred)

best_lambda = lambda_grid[np.argmin(train_mse)]
best_boost_hit = GradientBoostingRegressor(
    loss="squared_error",
    n_estimators=1000,
    learning_rate=best_lambda,
    random_state=1,
)
best_boost_hit.fit(xtrain, ytrain)

summary = pd.DataFrame(
    {
        "var": xtrain.columns,
        "rel.inf": 100 * best_boost_hit.feature_importances_ / best_boost_hit.feature_importances_.sum(),
    }
).sort_values("rel.inf", ascending=False)

print(summary)

File: task_equivalents/llm_translations/gpt/optimized/r/machine_learning_workflows/02_regression_workflow_feature_importance.R
library(readr)
library(dplyr)
library(gbm)
library(glmnet)

Hitters <- read_csv("data/csv/Hitters.csv", show_col_types = FALSE) |>
tidyr::drop_na()

drop_cols <- intersect(c("rownames", "Player", "Name"), names(Hitters))
if (length(drop_cols) > 0) {
Hitters <- Hitters |> select(-all_of(drop_cols))
}

Hitters <- Hitters |>
mutate(
League = recode(League, A = 0, N = 1),
NewLeague = recode(NewLeague, A = 0, N = 1),
Division = recode(Division, E = 0, W = 1),
logSal = log(Salary)
)

x <- Hitters |> select(-Salary, -logSal)
y <- Hitters$logSal

set.seed(1)
train_idx <- sample.int(nrow(x), size = 200)
xtrain <- x[train_idx, , drop = FALSE]
xtest <- x[-train_idx, , drop = FALSE]
ytrain <- y[train_idx]
ytest <- y[-train_idx]

lambda_grid <- seq(0.001, 0.201, length.out = 20)
train_mse <- numeric(length(lambda_grid))

xtrain_mat <- as.matrix(xtrain)
xtest_mat <- as.matrix(xtest)

for (i in seq_along(lambda_grid)) {
boost <- gbm(
formula = ytrain ~ .,
data = data.frame(ytrain = ytrain, xtrain),
distribution = "gaussian",
n.trees = 1000,
interaction.depth = 1,
shrinkage = lambda_grid[i],
bag.fraction = 0.5,
train.fraction = 1,
n.minobsinnode = 10,
verbose = FALSE
)
pred_train <- predict(boost, newdata = xtrain, n.trees = 1000)
train_mse[i] <- mean((ytrain - pred_train)^2)
}

lasso <- cv.glmnet(xtrain_mat, ytrain, alpha = 1, nfolds = 10)
lasso_pred <- predict(lasso, newx = xtest_mat, s = "lambda.min")
lasso_mse <- mean((ytest - lasso_pred)^2)

best_lambda <- lambda_grid[which.min(train_mse)]
best_boost <- gbm(
formula = ytrain ~ .,
data = data.frame(ytrain = ytrain, xtrain),
distribution = "gaussian",
n.trees = 1000,
interaction.depth = 1,
shrinkage = best_lambda,
bag.fraction = 0.5,
train.fraction = 1,
n.minobsinnode = 10,
verbose = FALSE
)

feature_importance <- summary(best_boost, n.trees = 1000, plotit = FALSE)
feature_importance <- setNames(feature_importance$rel.inf, feature_importance$var)
feature_importance <- sort(feature_importance, decreasing = TRUE)
print(feature_importance)